electrum/lnrouter.py

```diff
@@ -161,12 +167,251 @@ def is_fee_sane(fee_msat: int, *, payment_amount_msat: int) -> bool:
+        was chosen such that the penalty will be able to compete with the regular
+        base and relative fees.
+        """
+        hint = self.get_hint(node_from, node_to, channel_id)
+        can_send = hint.can_send(node_from < node_to)
+        cannot_send = hint.cannot_send(node_from < node_to)
+
+        # if we know nothing about the channel, return a default penalty
+        if (can_send, cannot_send) == (None, None):
+            return fee_for_edge_msat(amount, DEFAULT_PENALTY_BASE_MSAT, DEFAULT_PENALTY_PROPORTIONAL_MILLIONTH)
+        # next cases are with half information
+        elif can_send and not cannot_send:
+            if amount <= can_send:
+                return 0
+            else:
+                return fee_for_edge_msat(amount, DEFAULT_PENALTY_BASE_MSAT, DEFAULT_PENALTY_PROPORTIONAL_MILLIONTH)
+        elif not can_send and cannot_send:
+            if amount >= cannot_send:
+                return inf
+            else:
+                return fee_for_edge_msat(amount, DEFAULT_PENALTY_BASE_MSAT, DEFAULT_PENALTY_PROPORTIONAL_MILLIONTH)
+        # we know how much we can/cannot send
+        elif can_send and cannot_send:
+            if amount <= can_send:
+                return 0
+            elif amount < cannot_send:
+                return fee_for_edge_msat(amount, DEFAULT_PENALTY_BASE_MSAT, DEFAULT_PENALTY_PROPORTIONAL_MILLIONTH)
+            else:
+                return inf
+        return 0
+
+    @with_lock
+    def blacklist(self, node_from: bytes, node_to: bytes, channel_id: ShortChannelID):
+        hint_key = self._hint_key(node_from, node_to, channel_id)
+        hint = self._liquidity_hints.get(hint_key)
+        if not hint:
+            hint = self._liquidity_hints[hint_key] = LiquidityHint()
+        hint.blacklisted = True
+        now = int(time.time())
+        hint.timestamp = now
+
+    @with_lock
+    def get_blacklist(self) -> Set[ShortChannelID]:
+        now = int(time.time())
+        return set(k[2] for k, v in self._liquidity_hints.items() if now - v.timestamp < BLACKLIST_DURATION)
+
+    @with_lock
+    def clear_blacklist(self):
+        for k, v in self._liquidity_hints.items():
+            v.blacklisted = False
+
+    def __repr__(self):
+        string = "liquidity hints:\n"
+        if self._liquidity_hints:
+            for k, v in self._liquidity_hints.items():
+                string += f"{k}: {v}\n"
+        return string
+
 
 class LNPathFinder(Logger):
 
     def __init__(self, channel_db: ChannelDB):
         Logger.__init__(self)
         self.channel_db = channel_db
+        self.liquidity_hints = LiquidityHints()
+
+    def update_liquidity_hints(
+            self,
+            route: LNPaymentRoute,
+            amount_msat: int,
+            failing_channel: ShortChannelID=None
+    ):
+        # go through the route and record successes until the failing channel,
+        # for the failing channel, add a cannot_send liquidity hint
```

Comment:
maybe add a short fixme that the actual forwarded amounts are a bit different as the routing fees are ignored atm